contracts/margin/external/BucketLender/BucketLender.sol

```diff
@@ -0,0 +1,1129 @@
+     */
+    function marginCallOnBehalfOf(
+        address caller,
+        bytes32 positionId,
+        uint256 depositAmount
+    )
+        external
+        onlyMargin
+        onlyPosition(positionId)
+        returns (address)
+    {
+        require(
+            TRUSTED_MARGIN_CALLERS[caller],
+            "BucketLender#marginCallOnBehalfOf: Margin-caller must be trusted"
+        );
+        require(
+            depositAmount == 0, // disallows any deposit amount to cancel the margin-call
+            "BucketLender#marginCallOnBehalfOf: Deposit amount must be zero"
+        );
+
+        return address(this);
+    }
+
+    /**
+     * Function a contract must implement in order to let other addresses call cancelMarginCall().
+     *
+     * @param  canceler    Address of the caller of the cancelMarginCall function
+     * @param  positionId  Unique ID of the position
+     * @return             This address to accept, a different address to ask that contract
+     */
+    function cancelMarginCallOnBehalfOf(
+        address canceler,
+        bytes32 positionId
+    )
+        external
+        onlyMargin
+        onlyPosition(positionId)
+        returns (address)
+    {
+        require(
+            TRUSTED_MARGIN_CALLERS[canceler],
+            "BucketLender#cancelMarginCallOnBehalfOf: Margin-call-canceler must be trusted"
+        );
+
+        return address(this);
+    }
+
+    /**
+     * Function a contract must implement in order to let other addresses call
+     * forceRecoverCollateral().
+     *
+     *  param  recoverer   Address of the caller of the forceRecoverCollateral() function
+     * @param  positionId  Unique ID of the position
+     * @param  recipient   Address to send the recovered tokens to
+     * @return             This address to accept, a different address to ask that contract
+     */
+    function forceRecoverCollateralOnBehalfOf(
+        address /* recoverer */,
+        bytes32 positionId,
+        address recipient
+    )
+        external
+        onlyMargin
+        onlyPosition(positionId)
+        returns (address)
+    {
+        require(
+            recipient == address(this),
+            "BucketLender#forceRecoverCollateralOnBehalfOf: Recipient must be this contract"
+        );
+
+        rebalanceBuckets();
+
+        wasForceClosed = true;
+
+        return address(this);
+    }
+
+    // ============ Public State-Changing Functions ============
+
+    /**
+     * Allows users to deposit owedToken into this contract. Allowance must be set on this contract
+     * for "token" in at least the amount "amount".
+     *
+     * @param  beneficiary  The account that will be entitled to this depoit
+     * @param  amount       The amount of owedToken to deposit
+     * @return              The bucket number that was deposited into
+     */
+    function deposit(
+        address beneficiary,
+        uint256 amount
+    )
+        external
```

Comment:
[critical] You need to make like every single one of these public functions `nonReentrant`

Reply:
done